keps/0007-command-asserts.md

```diff
@@ -0,0 +1,103 @@
+---
+kep-number: 7
+short-desc: Assertions for CLI commands
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2020-10-28
+last-updated: 2020-10-28
+status: provisional
+---
+
+# Assertions for CLI commands
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [User Story 1](#user-story-1)
+    - [User Story 2](#user-story-2)
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+By asserting CLI commands, kuttl can support many additional test scenarios that examing the internal state of an operator.
+
+## Motivation
+
+Operators and custom resources can be used to deploy complex applications. For example, KUDO provides operators for Apache Cassandra and Apache Kafka. Deploying custom resources of these operators may not always result in changed Kubernetes resources but also in changes that aren't surfaced by Kubernetes resources but are internal to the operator. Asserting these changes only works through accessing service APIs or running commands and parsing their output. kuttl should provide tools to assert such output.
+
+### Goals
+
+- Assert that commands examining the internal state of an operator succeed or fail
+- Keep eventual consistency when asserting commands
+
+### Non-Goals
+
+- Introduce "flows" to determine when to run commands. I.e. to run commands once before or after asserting resources
+
+## Proposal
+
+Introduce commands as part of `TestAssert`. These commands are run at a regular interval like resource assertions until they either succeed or a timeout of the assert is reached. If any of the commands fail, the assert is considered to have failed. A command's return value determines if a command failed or not.
+
+As a start, complex commands that parse output can be described using shell environments. At a later stage we can implement additional tools to simplify common tasks. E.g. to search with a regular expression in command output.
```

Comment:
the proposal should include:

* the difference from existing asserts (which is that they are fired until true and no more... other asserts will fire repeated until all asserts are true)
* we should define the output of a command assert failure (currently we have verbose diff on declarative asserts... this will clearly be different)
* we should define how the report output will be different and what it looks like
* we should define what defines a failure and what defines a success...  for example 0 == success and anything else is a failure?

Reply:
we need to be clear if commands are part of the timeout and how that applies
it also needs to be clear how to change the timeout for a specific command assert

(I'm thinking here... what if someone does a `tail -f` or `kubectl logs -f`... eventually this should timeout)

Reply:
I added details around command output and exit codes to the proposal.